Migrate proof of time to use chain height

rahulksnv opened the pull request:
We currently use timestamp as the slot number in the proofs. This is changed to use block chain style height in the slot number.

Main changes:
1. Since the genesis slot issue is no longer applicable, the workarounds are cleaned up. Time keeper, node client initialization becomes much simpler now. The pre digest also clean up.
2. Consensus uses PoT chain as the slot mechanism, to drive block proposal. The pre digest slot now keeps the PoT slot number
3. Other clean ups, more checks in proof validation path.

https://github.com/subspace/subspace/issues/1827

### Code contributor checklist:
* [X] I have read, understood and followed [contributing guide](https://github.com/subspace/subspace/blob/main/CONTRIBUTING.md)

rahulksnv committed:
Change proof slot number to use PoT chain height.

Other changes:
1. Clean up the work arounds, initialization that was needed
   to work around the genesis slot issue. Since the PoT
   chain now uses the 0-based height, then genesis slot
   issue is no longer applicable.
2. The pre digest slot is changed to carry the PoT chain
   height as well.
3. Corresponding to pre digest simplifications.
4. More checks in the proof validation path.

nazar-pc commented on crates/pallet-subspace/src/lib.rs:
```diff
@@ -1579,6 +1582,19 @@ impl<T: Config> OnTimestampSet<T::Moment> for Pallet<T> {
     }
 }
 
+#[cfg(feature = "pot")]
+impl<T: Config> OnTimestampSet<T::Moment> for Pallet<T> {
+    fn on_timestamp_set(_moment: T::Moment) {
+        let slot_duration = Self::slot_duration();
+        assert!(
+            !slot_duration.is_zero(),
+            "Subspace slot duration cannot be zero."
+        );
+
+        // TODO: more checks
+    }
+}
```
Why is this necessary at all? I do not think we need to check anything about timestamp in consensus anymore.

nazar-pc reviewed (commented):
I scrolled through it, this PR seems to introduce many of the same or similar changes already present in https://github.com/subspace/subspace/pull/1854.

Some changes are probably better (I was only going to get something functional), others aren't (no restart support anywhere even potentially).

Slot notion here seems to be a hack specific to PoT more than anything, I believe we need to go from first principles and bootstrap correctly with PoT slots rather than having two notions of slots concurrently.

Seems like we worked on almost exactly the same changes here that will be hard to merge together :confused:

rahulksnv committed:
Add support for receiving notifications when proofs are added.

1. When a proof is added to the chain, notfication is sent with the proof.
2. Change consensus slot worker to listen to these events, instead of
   querying for the latest tip (push vs pull model). An equivalent of
   start_slot_worker()/next_slot() is implemented. The notification
   acts like a clock tick to trigger block building.
3. Other clean ups
